### PR TITLE
Fix #3806: return a flag to indicate whether the response is processed

### LINKED ISSUE #3806: The error page handler didn't process correctly in proxy
Hi, I’m using Jetty 9.4.18.v20190429 to implement the reverse proxy function.
but it has a problem in Chrome 73 when customizing the error page.

Case:
Chrome added a new header application/signed-exchange;v=b3 and this is not a text/ style header since version 73. and this cause a different behavior with the firefox.
On chrome it’s (note the content-length is 0)
image
image
On firefox:
image

after digging more:
Code analysis:

`org.eclipse.jetty.server.handler.ErrorHandler#generateAcceptableResponse(org.eclipse.jetty.server.Request, javax.servlet.http.HttpServletRequest, javax.servlet.http.HttpServletResponse, int, java.lang.String)`
```
        List<String> acceptable=baseRequest.getHttpFields().getQualityCSV(HttpHeader.ACCEPT, QuotedQualityCSV.MOST_SPECIFIC_MIME_ORDERING);
        
        if (acceptable.isEmpty() && !baseRequest.getHttpFields().contains(HttpHeader.ACCEPT))
        {
            generateAcceptableResponse(baseRequest,request,response,code,message,MimeTypes.Type.TEXT_HTML.asString());
        }
        else
        {
            for (String mimeType:acceptable)
            {
                generateAcceptableResponse(baseRequest,request,response,code,message,mimeType);
                if (baseRequest.isHandled())
                    break;
            }
        }
        baseRequest.setHandled(true);
        baseRequest.getResponse().closeOutput();
```
this method has a loop on the accept headers to process and generate a readable message.
but this will NOT work because the 
`baseRequest.isHandled()  is always true when entering this code`.

### PR DESCRIPTION
**Pull request by gaoxingliang**

### CODE REVIEW

**Comment by gaoxingliang** [comment]
Hi @joakime 
in your comment about the remove of  `baseRequest.isHandled()`
if we don’t remove this line, it will not fix. because the request has been marked as handled in previous code.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java [hunk 1]
```diff
@@ -46,6 +32,20 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java]: leave imports in order

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java [hunk 2]
```diff
@@ -167,9 +167,9 @@ protected void generateAcceptableResponse(Request baseRequest, HttpServletReques
         {
             for (String mimeType : acceptable)
             {
-                generateAcceptableResponse(baseRequest, request, response, code, message, mimeType);
-                if (baseRequest.isHandled())
+                if (generateAcceptableResponse(baseRequest, request, response, code, message, mimeType)) {
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java]: please apply coding style so that { is on a new line

**Comment by gaoxingliang** [comment]
@gregw 
I uploaded a small test code for you to re-produce it here:
https://github.com/gaoxingliang/goodutils/blob/master/proxydemeo/JettyIssue3806.MD

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java (hunk 1 above)
Comment (gaoxingliang) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java]: Fixed

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java (hunk 2 above)
Comment (gaoxingliang) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java]: Fixed

**Comment by gregw** [comment]
@gaoxingliang  I still don't understand why you want to do this?
Your example has:
```java
        protected void generateAcceptableResponse(Request baseRequest, HttpServletRequest request, HttpServletResponse response, int code,
                                                  String message, String mimeType) throws IOException {
            super.generateAcceptableResponse(baseRequest, request, response, code, message, mimeType);
            /**
             * We found the chrome added a Accept type since chrome 73.
             * and this cause the error message is not showing eg when a port is redirected from 80 to 443.
             * here we ignore the MIME type is okay. we really do return a HTML page
             */
        }
```
So can't you just change that to 
```java
        protected void generateAcceptableResponse(Request baseRequest, HttpServletRequest request, HttpServletResponse response, int code,
                                                  String message, String mimeType) throws IOException {

             /* ignore the MIME type is okay. we really do return a HTML page */
            baseRequest.getHttpFields().remove(HttpHeader.ACCEPT);
            super.generateAcceptableResponse(baseRequest, request, response, code, message, mimeType);
        }
```

Or
```java
        protected void generateAcceptableResponse(Request baseRequest, HttpServletRequest request, HttpServletResponse response, int code,
                                                  String message, String mimeType) throws IOException {

             /* Always accept HTML error page */
            baseRequest.getHttpFields().add(HttpHeader.ACCEPT, MimeTypes.Type.TEXT_HTML.asString());
            super.generateAcceptableResponse(baseRequest, request, response, code, message, mimeType);
        }
```

**Comment by gaoxingliang** [comment]
@gregw 
The workaround you provided will not work. because it’s still using the input mimeType.
below code will work around:
```
 super.generateAcceptableResponse(baseRequest, request, response, code, message, MimeTypes.Type.TEXT_HTML.asString());     
```

- Why we do this?

We are implementing a reverse proxy and we want to wrap and return our own error page instead of default error page with power by and anything else.

- Why I don’t think this is a good idea?
in code `org.eclipse.jetty.server.handler.ErrorHandler#generateAcceptableResponse(org.eclipse.jetty.server.Request, javax.servlet.http.HttpServletRequest, javax.servlet.http.HttpServletResponse, int, java.lang.String)`
```
            for (String mimeType:acceptable)
            {
                // line 3
                generateAcceptableResponse(baseRequest,request,response,code,message,mimeType);
                if (baseRequest.isHandled())
                    break;
            }
```
it try to use `baseRequest.isHandled` to decide whether to continue to try more mimeType.  but the baseRequest.isHandled always return true no matter whether it’s set handled in the method line 3.

**Comment by gregw** [comment]
I'm sorry but I'm just not understanding your use case. Can you give me an example program that demonstrates the problem rather than illustrates your proposed solution?

**Comment by gaoxingliang** [comment]
@gregw 
Here is the demo code
https://github.com/gaoxingliang/goodutils/blob/master/proxydemeo/JettyIssue3806.MD

**Comment by gregw** [comment]
Ah yes, my fix wont work, but it is easy to fix the fix the next layer up:
```java
        @Override
        protected void generateAcceptableResponse(Request baseRequest, HttpServletRequest request, HttpServletResponse response, int code, String message) throws IOException {
             /* Always accept HTML error page */
            baseRequest.getHttpFields().add(HttpHeader.ACCEPT, MimeTypes.Type.TEXT_HTML.asString());
            super.generateAcceptableResponse(baseRequest, request, response, code, message);
        }
```

Calling the normal `generateAcceptableResponse` implementation that takes a mimeType does not always return true.  It only handles:
```java
            case "text/html":
            case "text/*":
            case "*/*":
```
and all other mimeTypes are noops (unless somebody has extended the method to handle other types.... which we know they do, so we can't change the signature of this method).

Finally, you are free to implement `ErrorHandler.doError` however you want.  There is no requirement to use the provided convenience methods like `generateAcceptableResponse`.  If they are not convenient, then replace them with your own handling.  This class is extended by other users, so we are constrained in how we change the signature and behaviour.  We might change it to be more convenient for your use-case, but in the process break many other existing users!

**Comment by gaoxingliang** [comment]
yes. it's easy to work around. but it's kind of confusing. 
```
for (String mimeType:acceptable)
            {
                // line 3
                generateAcceptableResponse(baseRequest,request,response,code,message,mimeType);
                if (baseRequest.isHandled())
                    break;
            }
```
it's here.   Check my previous comment. just for your information.

**Pull request closed without merging**